### DISCUSSION TITLE
EntityLinker: Cannot get dimension 'nO' for model 'transformer-listener>>trfs2arrays

### DISCUSSION THREAD

**Discussion by adamkgoldfarb** [explosion/spaCy#7884] [Help: Coding & Implementations] [answered]
Hi @svlandeg -- I'm attempting to train an entity linker with a transformer component as a drop-in replacement for the tok2vec layer. I am getting the error:

`ValueError: Cannot get dimension 'nO' for model 'transformer-listener>>trfs2arrays': value unset`

I checked through [this issue](https://github.com/explosion/spaCy/discussions/6789) but I don't think it's directly related-- happy to be wrong!

I'm likely missing something in the config, but I've tried setting nO to 768 rather than null where it's noted in the config and am still getting that error. Here's the config-- any insight would be super appreciated, and thanks in advance!

----

[paths]
train = "training/lea_nel_train/lea_nel_train.spacy"
dev = "training/lea_nel_train/lea_nel_dev.spacy"
vectors = null
init_tok2vec = null
raw_text = "corpus/rfp_free_responses.jsonl"
kb = "knowledge_bases/lea_entities_roberta_trf"
base_nlp = "models/lea_roberta_base_trf/model-best"

[system]
gpu_allocator = "pytorch"
seed = 30

[nlp]
lang = "en"
pipeline = ["transformer","ner","entity_linker"]
batch_size = 128
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.entity_linker]
factory = "entity_linker"
entity_vector_length = 768
get_candidates = {"@misc":"spacy.CandidateGenerator.v1"}
incl_context = true
incl_prior = true
n_sents = 1
labels_discard = []

[components.entity_linker.model]
@architectures = "spacy.EntityLinker.v1"
nO = null

[components.entity_linker.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"

[components.ner]
source = ${paths.base_nlp}

[components.transformer]
source = ${paths.base_nlp}

[corpora]

[corpora.train]
@readers = "MyCorpus.v1"
file = ${paths.train}

[corpora.dev]
@readers = "MyCorpus.v1"
file = ${paths.dev}

[training]
train_corpus = "corpora.train"
dev_corpus = "corpora.dev"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.2
patience = 10000
eval_frequency = 200
accumulate_gradient = 2
max_epochs = 0
max_steps = 600
frozen_components = ["transformer","ner"]
before_to_disk = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = true

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
nel_micro_p = 0.0
nel_micro_r = 0.0
nel_micro_f = 1.0

[pretraining]

[optimizer]
@optimizers = "Adam.v1"
learn_rate = 0.001
beta1 = 0.9
beta2 = 0.999
L2 = 0.0
eps = 0.00000001
grad_clip = 1.0
L2_is_weight_decay = true
use_averages = true

[initialize]
vectors = null
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.components.entity_linker]

[initialize.components.entity_linker.kb_loader]
@misc = "spacy.KBFromFile.v1"
kb_path = ${paths.kb}

[initialize.tokenizer]

**Comment by svlandeg** [comment]
Hi! 

Could you check whether this still happens if you don't put the `transformer` in the frozen list? If it does still happen - can you paste the full stack trace?

A related discussion & explanation by Adriane is here: https://github.com/explosion/spaCy/discussions/7811#discussioncomment-639822. In short, 3.0 has some "quircks" when freezing some layers of the listener pattern and training others, like in this case, so until 3.1 we'll have to find a way around that.

**Reply by adamkgoldfarb** [reply]
Hello again-- thanks for linking that thread. That sounds exactly like the issue I'm having, but when I take the transformer out of the frozen list and just leave the NER component, the error still occurs. The stacktrace is below. 

I have some experimentation to do with a static vectors tok2vec and will try with that as well while I wait on v3.1, though in that case it might make more sense to include two tok2vec layers-- I assume that I can refer to the same static vectors but train different tok2vec components. 

The real benefit in my use case to using transformers instead of the standard tok2vec component seems to be (intermediate user here) the token context I'm getting passed forward to NER-- but if it turns out that I've been training the transformer layer during NER training (as it seems we train the tok2vec layer) then I might need to reevaluate my training strategy!

Thanks again-- let me know if you need more info here.

Adam

=============================== train_lea_nel ===============================
Running command: /Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/bin/python -m spacy train configs/config_trf_nel.cfg --output training/lea_nel_train -c scripts/nel_custom_functions.py
ℹ Using CPU

=========================== Initializing pipeline ===========================
Traceback (most recent call last):
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/__main__.py", line 4, in <module>
    setup_cli()
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/cli/_util.py", line 69, in setup_cli
    command(prog_name=COMMAND)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/typer/main.py", line 497, in wrapper
    return callback(**use_params)  # type: ignore
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/cli/train.py", line 56, in train_cli
    nlp = init_nlp(config, use_gpu=use_gpu)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/training/initialize.py", line 39, in init_nlp
    nlp = load_model_from_config(raw_config, auto_fill=True)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/util.py", line 426, in load_model_from_config
    nlp = lang_cls.from_config(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/language.py", line 1662, in from_config
    nlp.add_pipe(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/language.py", line 768, in add_pipe
    pipe_component = self.create_pipe(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/language.py", line 659, in create_pipe
    resolved = registry.resolve(cfg, validate=validate)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/config.py", line 722, in resolve
    resolved, _ = cls._make(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/config.py", line 771, in _make
    filled, _, resolved = cls._fill(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/config.py", line 826, in _fill
    filled[key], validation[v_key], final[key] = cls._fill(
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/config.py", line 843, in _fill
    getter_result = getter(*args, **kwargs)
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/spacy/ml/models/entity_linker.py", line 14, in build_nel_encoder
    token_width = tok2vec.get_dim("nO")
  File "/Users/adamgoldfarb/opt/miniconda3/envs/bmgf-k12-pbd/lib/python3.9/site-packages/thinc/model.py", line 175, in get_dim
    raise ValueError(err)
ValueError: Cannot get dimension 'nO' for model 'transformer-listener>>trfs2arrays': value unset

**Reply by svlandeg** [reply]
Thanks for testing and the details, let me look into this in a bit more detail.

**Reply by svlandeg** [reply]
Ok, hm, there's a few issues here. First and foremost I think the NEL is in need of a refactor sometime soon, as it wasn't designed with a transformer-based architecture in mind, and there is currently too strong a connection between the vectors in the `KnowledgeBase` and the `entity_linker` architecture.

Anyway. Just to double check, how did you create your `KnowledgeBase` - what are the dimensions of the entity vectors in it?

**Reply by adamkgoldfarb** [reply]
Sorry, just saw this-- I created the vectors in the kb by making a doc object on description text with a model with pipeline components: ["transformer", "ner"]. I used user hooks per spacy v2 to recover 768d vectors from the tensors. Let me know if you need more info here!

Thanks again.

**Reply by svlandeg** [reply]
No, that sounds good :-)

**Reply by moyid** [reply]
Hi, @adamkgoldfarb @svlandeg I'm working on a very similar pipeline, would you mind expanding on how you used user hooks to get the vectors for the kb?

**Reply by adamkgoldfarb** [reply]
[This article](https://applied-language-technology.readthedocs.io/en/latest/notebooks/part_iii/04_embeddings_continued.html#contextual-embeddings-from-transformers) made it all clear for me and really spells it out toward the end! You'll need to add the tensor2attr pipe to your nlp and then you can access the tensor encodings rather than the static vector encodings, because you will have overridden the .vector attr with trf tensors on the doc instance.

**Reply by moyid** [reply]
thank you @adamkgoldfarb ! That was very helpful!

**Reply by moyid** [reply]
sadly this did not solve my issue. I think I got the vector right, but I am still getting the same error. @adamkgoldfarb @svlandeg   If you don't mind taking a look at my issue - https://github.com/explosion/spaCy/discussions/9310 , it is very similar to this setup. I'd really appreciate it!

**Reply by adamkgoldfarb** [reply]
Going to keep my reply here because I'm afraid I'm out of my depth-- one of the differences I see between our setups is the sentencizer. Are you sure you need that component? It might throw a wrench in the trf processing. I might try removing it. But I need to emphasize that's a guess!

**Reply by moyid** [reply]
Thanks, I tried it without sentencizer as well, and still got the same error :/

**Comment by svlandeg** [comment, marked as answer]
(reposting here to be able to select as answer): Assuming your vectors in the KB are of length 768, this fix should hopefully resolve your problem: #7930

**Reply by adamkgoldfarb** [reply]
Thank you so much for your attention to this! Looking forward to trying it out! I'll monitor the pull request and test when merged. 

Much, much appreciated!

**Reply by svlandeg** [reply]
Merged :-)
